### PR TITLE
lpc47m157: Fix serial

### LINKED ISSUE #86: Serial is broken
QEMU 5.1 both 32bit and 64bit are able to use the ports correctly without crashing - I assume it's something to do with the custom backend? Not sure. Issue happens with both elevated and not.

In the gif I was using a retail kernel however it makes no difference using a real debug or hacked kernel.

Launched with: `.\xemu.exe -device lpc47m157 -serial COM1`
Version: `2702f45`

```
ERROR:hw/core/qdev.c:403:qdev_assert_realized_properly: assertion failed: (dev->realized)
```

output

### PR DESCRIPTION
**Pull request by haxar**

### CODE REVIEW

**Comment by GXTX** [comment]
With this patch xemu no longer crashes *before* innit however on Windows it locks up if you attempt to attach WinDbg.

**Comment by haxar** [comment]
The `configuration_mode` asserts are removed in this PR since Linux touches `ENTER_CONFIG_KEY` more than once & `configuration_mode` doesn't seem to be used elsewhere.

File: hw/xbox/lpc47m157.c [hunk 1]
```diff
@@ -113,10 +113,8 @@ static void lpc47m157_io_write(void *opaque, hwaddr addr, uint64_t val,
     if (addr == 0) {
         /* INDEX_PORT */
         if (val == ENTER_CONFIG_KEY) {
-            assert(!s->configuration_mode);
             s->configuration_mode = true;
         } else if (val == EXIT_CONFIG_KEY) {
-            assert(s->configuration_mode);
```
Comment (binarymaster) [inline hw/xbox/lpc47m157.c:123]: This basically supersedes my old stalled PR at XQEMU https://github.com/xqemu/xqemu/pull/282

**Comment by haxar** [comment]
This latest force-push saves the state of the device registers & reconfigures the devices (serial) to the saved state when using snapshots. Works with Cromwell & xbox-linux.

File: hw/xbox/lpc47m157.c [hunk 2]
```diff
@@ -51,204 +56,298 @@
...
+
 static void lpc47m157_class_init(ObjectClass *klass, void *data)
 {
     DeviceClass *dc = DEVICE_CLASS(klass);
+    ISADeviceClass *isa = ISA_DEVICE_CLASS(klass);
 
     dc->realize = lpc47m157_realize;
     dc->vmsd = &vmstate_lpc47m157;
-    //dc->reset = pc87312_reset;
+    isa->build_aml = lpc47m157_build_aml;
```
Comment (mborgerson) [inline hw/xbox/lpc47m157.c]: Xbox doesn't support ACPI, we don't need to build AML

File: hw/xbox/lpc47m157.c [hunk 3]
```diff
@@ -51,204 +56,298 @@
...
+                    | dev[CONFIG_DEVICE_BASE_ADDRESS_LOW];
+        irq = dev[CONFIG_DEVICE_INTERRUPT] & 0x0f;
+        if ((!dev[CONFIG_DEVICE_ACTIVATE]
+            || serial->iobase != iobase
+            || serial->irq != irq)
+            && serial->active) {
+            isa_unregister_ioport(NULL, &ss->io);
+            memory_region_destroy(&ss->io);
+            ss->irq = NULL;
+
```
Comment (mborgerson) [inline hw/xbox/lpc47m157.c]: nit: blankline

File: hw/xbox/lpc47m157.c [hunk 4]
```diff
@@ -51,204 +56,298 @@
...
+            serial->active = false;
+            DPRINTF("lpc47m157 COM%d disabled @ iobase=0x%x irq=%u\n",
+                i + 1,
+                serial->iobase,
+                serial->irq);
+        }
+        if (dev[CONFIG_DEVICE_ACTIVATE]
```
Comment (mborgerson) [inline hw/xbox/lpc47m157.c]: nit: wrapping?

**Comment by mborgerson** [comment]
Thanks for working on fixing this! In the future, please avoid mixing formatting changes with bug fixes. It makes it more cumbersome to review the actual logic changes when the whole file changes.

File: hw/xbox/lpc47m157.c [hunk 5]
```diff
@@ -51,204 +56,298 @@
...
 } LPC47M157State;
 
-#define LPC47M157_DEVICE(obj) \
-    OBJECT_CHECK(LPC47M157State, (obj), "lpc47m157")
+typedef struct ISALPC47M157State {
+    ISADevice parent_obj;
 
-static void update_devices(LPC47M157State *s)
+    bool sysopt;
+    uint32_t iobase;
```
Comment (mborgerson) [inline hw/xbox/lpc47m157.c]: nit: I/O space is 16b addressable, let's use a `uint16_t` here

File: hw/xbox/lpc47m157.c [hunk 6]
```diff
@@ -51,204 +56,298 @@
...
     uint32_t val = 0;
 
-    if (addr == 0) {
-        /* INDEX_PORT */
-    } else if (addr == 1) {
-        /* DATA_PORT */
-        if (s->selected_reg < MAX_CONFIG_REG) {
-            val = s->config_regs[s->selected_reg];
+    if (s->configuration_mode) {
+        if (addr == INDEX_PORT) {
```
Comment (mborgerson) [inline hw/xbox/lpc47m157.c]: nit: Drop empty condition

File: hw/xbox/lpc47m157.c [hunk 7]
```diff
@@ -51,204 +56,298 @@
...
-
+    /* update serial devices */
+    for (i = 0; i < ARRAY_SIZE(s->serial); i++) {
+        serial = &s->serial[i];
+        ss = &serial->state;
+        dev = s->device_regs[DEVICE_SERIAL_PORT_1 + i];
+        iobase = (dev[CONFIG_DEVICE_BASE_ADDRESS_HIGH] << 8)
+                    | dev[CONFIG_DEVICE_BASE_ADDRESS_LOW];
+        irq = dev[CONFIG_DEVICE_INTERRUPT] & 0x0f;
+        if ((!dev[CONFIG_DEVICE_ACTIVATE]
```
Comment (mborgerson) [inline hw/xbox/lpc47m157.c]: nit: Suggest a slightly more readable, version with `serial->active` check first to short-circuit:

```c
        if (serial->active &&
            (!dev[CONFIG_DEVICE_ACTIVATE]
             || serial->iobase != iobase
             || serial->irq != irq)) {
```

File: hw/xbox/lpc47m157.c [hunk 8]
```diff
@@ -51,204 +56,298 @@
...
+
+            serial->active = false;
+            DPRINTF("lpc47m157 COM%d disabled @ iobase=0x%x irq=%u\n",
+                i + 1,
+                serial->iobase,
+                serial->irq);
+        }
+        if (dev[CONFIG_DEVICE_ACTIVATE]
+            && !serial->active) {
+            ss->irq = (irq != 0) ? isa_get_irq(&isa->parent_obj, irq) : NULL;
```
Comment (mborgerson) [inline hw/xbox/lpc47m157.c]: Unsure about this change. Can you clarify whether `isa_init_irq` should still be invoked

File: hw/xbox/lpc47m157.c (hunk 4 above)
Comment (haxar) [reply hw/xbox/lpc47m157.c:129]: This is intentional to reset/reactivate the serial device state if `iobase` or `irq` changes after loading from a snapshot via `lpc47m157_post_load()`.

File: hw/xbox/lpc47m157.c (hunk 8 above)
Comment (haxar) [reply hw/xbox/lpc47m157.c]: I could not find an equivalent `isa_deinit_irq()` when deactivating & `isa_init_irq()` doesn't seem to do anything critical other than to keep track of the requested IRQ number in the ISA device struct & could only be called 2 times. I figured `dev->nirqs++` might hit its assert in `isa_init_irq()` if called more than 2 times.

Comment (haxar) [reply hw/xbox/lpc47m157.c]: I'm able to confirm that this is indeed what happens:
```
xemu: hw/isa/isa-bus.c:96: isa_init_irq: Assertion `dev->nirqs < ARRAY_SIZE(dev->isairq)' failed.
```

**Comment by mborgerson** [comment]
Thanks for working on this, and for your patience while I review it. Apologies for the long delay.

I've rebased this branch, tweaked a commit message, and tested this as working on my machine. Works great, much appreciated!

**Pull request merged**